Px.Utils/Operations/MatrixFunctionExtensions.cs

```diff
@@ -0,0 +1,111 @@
+﻿using Px.Utils.Models;
+using Px.Utils.Models.Metadata;
+using Px.Utils.Models.Metadata.Dimensions;
+using Px.Utils.Models.Metadata.ExtensionMethods;
+using Px.Utils.PxFile.Data;
+
+namespace Px.Utils.Operations
+{
+    /// <summary>
+    /// Base class for matrix functions, defines some operation that can be applied to a matrix, such as a transformation or a sum
+    /// </summary>
+    /// <typeparam name="TData">Type of the data values in the matrix</typeparam>
+    public static class MatrixFunctionExtensions
+    {
+        /// <summary>
+        /// Applies the function to the input matrix, returning a new matrix with the result
+        /// </summary>
+        /// <param name="input">inoput matrix</param>
+        /// <returns>a new matrix with the result of the operation</returns>
+        public static Matrix<TData> ApplyOverDimension<TData>(
+            this Matrix<TData> input, DimensionValue newValue, IDimensionMap sourceMap, Func<TData, TData, TData> func, TData functionIdentity, int valueIndex = -1){
+            MatrixMetadata newMeta = valueIndex == -1
+                ? CopyMetaAndAddValue(input.Metadata, sourceMap.Code, newValue)
+                : CopyMetaAndInsertValue(input.Metadata, sourceMap.Code, newValue, valueIndex);
+
+            // Compute sum values to the output matrix
+
+            IMatrixMap sumOnlyMap = newMeta.CollapseDimension(sourceMap.Code, newValue.Code);
+            TData[] outData = new TData[input.Metadata.GetSize() + sumOnlyMap.GetSize()];
+            
+            // Initialize the output matrix with the identity value
+            for (int i = 0; i < outData.Length; i++) outData[i] = functionIdentity; 
+
+            for (int i = 0; i < sourceMap.ValueCodes.Count; i++)
+            {
+                IMatrixMap sumSubMap = input.Metadata.CollapseDimension(sourceMap.Code, sourceMap.ValueCodes[i]);
+                DataIndexer source = new(input.Metadata, sumSubMap);
+                DataIndexer target = new(newMeta, sumOnlyMap);
+
+                do outData[target.CurrentIndex] = func(outData[target.CurrentIndex], input.Data[source.CurrentIndex]);
+                while (source.Next() && target.Next());
+            }
+
+            // Copy the original data to the output matrix
+
+            DataIndexer original = new(newMeta, input.Metadata);
+            int ogIndex = 0;
+            do outData[original.CurrentIndex] = input.Data[ogIndex++];
+            while (original.Next());
+
+            return new Matrix<TData>(newMeta, outData);
+        }
+
```

Comment:
Summaries for this and ApplyRelative below?

Reply:
Added the missing documentation.